core/search/service.py

```diff
@@ -0,0 +1,330 @@
+
+    def index_submit_documents(
+        self, pointer: str, documents: Iterable[dict]
+    ) -> List[SearchServiceFailedDocument]:
+        self._logger.info(f"submitting documents to index {pointer}")
+
+        # Specifically override the target in all documents to the target pointer
+        # Add a hard requirement that the target be an alias (this prevents documents from implicitly creating
+        # indexes).
+        for document in documents:
+            document["_index"] = pointer
+            document["_require_alias"] = True
+
+        # See: Sources for "streaming_bulk":
+        # https://github.com/opensearch-project/opensearch-py/blob/db972e615b9156b4e364091d6a893d64fb3ef4f3/opensearchpy/helpers/actions.py#L267
+        # The documentation is incredibly vague about what the function actually returns, but these
+        # parameters _should_ cause it to return a tuple containing the number of successfully documents
+        # and a list of documents that failed. Unfortunately, the type checker disagrees and the documentation
+        # gives no hint as to what an "int" might mean for errors.
+        (success_count, errors) = opensearchpy.helpers.bulk(
+            client=self._client,
+            actions=documents,
+            raise_on_error=False,
+            max_retries=3,
+            max_backoff=30,
+            yield_ok=False,
+        )
+
+        error_results: List[SearchServiceFailedDocument] = []
+        if isinstance(errors, list):
+            for error in errors:
+                error_results.append(SearchServiceFailedDocument.from_bulk_error(error))
+        else:
+            raise SearchServiceException(
+                f"Opensearch returned {errors} instead of a list of errors."
+            )
+
+        return error_results
+
+    def index_clear_documents(self, pointer: str):
+        self._client.delete_by_query(
+            index=pointer, body={"query": {"match_all": {}}}, wait_for_completion=True
+        )
+
+    def refresh(self):
+        self._logger.debug(f"waiting for indexes to become ready")
+        self._client.indices.refresh()
+
+    def write_pointer_set(self, base_name: str, revision: SearchSchemaRevision) -> None:
+        alias_name = self.write_pointer_name(base_name)
+        target_index = revision.name_for_index(base_name)
+        action = {
+            "actions": [
+                {"remove": {"index": "*", "alias": alias_name}},
+                {"add": {"index": target_index, "alias": alias_name}},
+            ]
+        }
+        self._logger.debug(f"setting write pointer {alias_name} to {target_index}")
+        self._client.indices.update_aliases(body=action)
+
+    def read_pointer(self, base_name: str) -> Optional[str]:
+        try:
+            result: dict = self._client.indices.get_alias(
+                name=self.read_pointer_name(base_name)
+            )
+            for name in result.keys():
+                if name.startswith(f"{base_name}-"):
```

Comment:
What is the difference between the 2 checks for write and read pointers wherein one is a regex and this is a "startswith"?

Reply:
See above. :arrow_up: